### PR TITLE
KAFKA-9483; Add Scala KStream#toTable to the Streams DSL

### PR DESCRIPTION
**Pull request by highluck**
Add Scala KStream#toTable to the Streams DSL
https://issues.apache.org/jira/browse/KAFKA-9483

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by mjsax — APPROVED** [review]
LGTM.

@vvcephei @bbejeck -- I am not fluent in Scala -- can you have a second look?

**Review by bbejeck — APPROVED** [review]
LGMT pending passing build

**Comment by bbejeck** [comment]
Retest this please.

**Comment by mjsax** [comment]
Java 11: `KStreamImplTest.shouldSupportTriggerMaterializedWithKTableFromKStream` (should be covered via #8027 -- did not create a ticket)

Java 8:
```
kafka.admin.DeleteConsumerGroupsTest.testDeleteCmdWithMixOfSuccessAndError
kafka.admin.ResetConsumerGroupOffsetTest.testResetOffsetsExportImportPlan
```

We can trigger retesting after John's review.

**Comment by mjsax** [comment]
I just realized, that Bill review and approved already...

Retest this please.

**Comment by mjsax** [comment]
Both tests failed. Result are not available any longer.

Retest this please.

**Comment by mjsax** [comment]
I need to cycle back to this quicker. Same as before. Both tests failed but result are not available any longer.

Retest this please.

**Comment by mjsax** [comment]
Java 8:
```
kafka.admin.DeleteConsumerGroupsTest.testDeleteCmdWithMixOfSuccessAndError
```
Java 11:
```
kafka.admin.DescribeConsumerGroupTest.testDescribeGroupWithShortInitializationTimeout
```

Java 8 test failure is the same as above. Retest this please.

**Comment by mjsax** [comment]
@bbejeck @vvcephei @guozhangwang The Scala part of the API change was not explicitly discussed in the KIP -- do you think this is an issue and we should try to include such a discussion for all API changes?

Also, this PR is technically an API change, hence, I am wondering if it would be ok to cherry-pick to `2.5` branch or not? The feature itself was merged on time -- I personally think it would be ok to cherry-pick, but would like to double check what you think?

**Comment by mjsax** [comment]
Java 8 timed out.
Java 11:
```
kafka.admin.ResetConsumerGroupOffsetTest.testResetOffsetsExportImportPlan
```

Retest this please.

**Comment by mjsax** [comment]
Java 8:
```
kafka.admin.DescribeConsumerGroupTest.testDescribeGroupMembersWithShortInitializationTimeout
```
Java 11:
```
 kafka.admin.DeleteConsumerGroupsTest > testDeleteCmdWithMixOfSuccessAndError
```

I guess we can merge this?

**Comment by highluck** [comment]
@mjsax 
Thank you for your care!
But, No problem in my local environment

File: streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/KStreamTest.scala [hunk 1]
```diff
@@ -312,4 +312,28 @@ class KStreamTest extends FlatSpec with Matchers with TestDriver {
...
   }
+
+  "join 2 KStreamToTables" should "join correctly records" in {
+    val builder = new StreamsBuilder()
+    val sourceTopic1 = "source1"
+    val sourceTopic2 = "source2"
+    val sinkTopic = "sink"
+
+    val table1 = builder.stream[String, String](sourceTopic1).toTable.groupBy((k, v) => (k, v)).count()
+    val table2 = builder.stream[String, String](sourceTopic2).toTable.groupBy((k, v) => (k, v)).count()
```
Comment (vvcephei) [inline streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/KStreamTest.scala]: Sorry to chime in so late on this. I looked at this test before and missed something, which I have just now noticed.

This test inserts an extra operation that defeats the purpose of the test. We should not do `groupBy().count()`, just the `toTable` on both streams, followed immediately by the join. The joiner can just concatenate both values and the verification should read the output as a map and ensure the final value is `"topic1value1topic2value1"`.

Adding the groupBy and count results in a repartition and stateful operation, potentially masking a failure to auto-materialize the toTable.

If you like, you can leave this test (there's nothing wrong with it _per se_) and just add a new test doing what I described.

**Review by vvcephei — CHANGES_REQUESTED** [review]
Hey all, really sorry about this, but I just now noticed a significant issue in the test.

Regarding @mjsax 's earlier question, I can't believe we missed the scala API again in the KIP... We should just update the KIP and send a note to the vote thread. Also, we should cherry-pick to 2.5.

**Comment by bbejeck** [comment]
>Regarding @mjsax 's earlier question, I can't believe we missed the scala API again in the KIP... We should just update the KIP and send a note to the vote thread. Also, we should cherry-pick to 2.5.

+1

**Comment by highluck** [comment]
@vvcephei 
thanks for review ! 
Updated the KIP and updated the test.

retest this please

**Review by vvcephei — APPROVED** [review]
Thanks @highluck

**Comment by vvcephei** [comment]
Ok to test.

**Comment by vvcephei** [comment]
Retest this, please.

**Comment by mjsax** [comment]
Test is green. Merging.

**Comment by mjsax** [comment]
Both runs passed. Merging.

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `2.5`.

Thanks for the PR @highluck!

**Pull request merged**